sympy/plotting/plot_implicit.py

```diff
@@ -38,6 +38,9 @@
 from sympy.utilities.iterables import flatten
 
 
+__doctest_requires__ = {'plot_implicit': ['matplotlib']}
+
+
 @doctest_depends_on(modules=('matplotlib',))
```

Comment:
We already have this decorator. Can we just modify it to update `__doctest_requires__`?

Reply:
My ultimate goal is that we use `pytest-doctestplus` and don't bother with the in-house test runner for doctest so I would rather configure this the way that is suggested for `pytest-doctestplus`. 

We could change it the other way around and have `doctest_depends_on` check `__doctest_requires__` but again I see this is as being only temporary so I'm not sure what value there is. In the end I would want to remove `doctest_depends_on` and have only `__doctest_requires__` or whatever features pytest-doctestplus uses.

Reply:
Currently I have:
```console
$ pytest --doctest-plus -nauto
...
FAILED sympy/physics/continuum_mechanics/beam.py::sympy.physics.continuum_mechanics.beam.Beam.draw
FAILED sympy/solvers/tests/test_simplex.py::test_lp - ImportError: cannot import name '_np' from 'sympy.solvers.inequalities' (/Users/enojb/work/dev/sympy/sympy/solvers/inequalities.py)
================================================================== 2 failed, 17098 passed, 353 skipped, 328 xfailed, 5 xpassed in 179.53s (0:02:59) ==================================================================
```
So that's 3 minutes to run all not slow tests and doctests using pytest-xdist by spreading the load over multiple cores. This is much better than `bin/doctest`.

The only thing missing is better markdown support in pytest-doctestplus:
https://github.com/scientific-python/pytest-doctestplus/issues/247

Reply:
The decorator is much better than the variable because it keeps the metadata next to the function. We could upstream it if that makes you feel better. But I honestly don't see a big deal having helper decorators and functions on top of test frameworks. Every large library does that. 

Does doctestplus support requiring all imports in every function? That was the main drawback the last time I checked.

Reply:
> Does doctestplus support requiring all imports in every function?

I think so. Feel free to check. I just tried:
```python
"""

>>> from sympy import cos

"""

def f(x):
    """
    >>> cos(0)
    1
    """
    return 1
```
And I get:
```console
$ pytest --doctest-plus t.py
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.12.2, pytest-8.1.1, pluggy-1.4.0
rootdir: /Users/enojb/work/dev/sympy
configfile: pyproject.toml
plugins: instafail-0.5.0, hypothesis-6.99.9, xdist-3.5.0, timeout-2.3.1, split-0.8.2, doctestplus-1.2.2.dev3+g84fd2e2.d20240402
collected 2 items

t.py .F                                                                                                                                                                                                        [100%]

====================================================================================================== FAILURES ======================================================================================================
___________________________________________________________________________________________________ [doctest] t.f ____________________________________________________________________________________________________
008
009     >>> cos(0)
UNEXPECTED EXCEPTION: NameError("name 'cos' is not defined")
Traceback (most recent call last):
  File "/Users/enojb/.pyenv/versions/3.12.2/lib/python3.12/doctest.py", line 1361, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest t.f[0]>", line 1, in <module>
NameError: name 'cos' is not defined
/Users/enojb/work/dev/sympy/t.py:9: UnexpectedException
============================================================================================== short test summary info ===============================================================================================
FAILED t.py::t.f
============================================================================================ 1 failed, 1 passed in 0.14s =============================================================================================
```

Reply:
In this case, you would need to check that `f` is not defined in the doctest unless it is imported. The default behavior of doctest is to assume the doctest runs in the module namespace of the module the function is defined in.

Reply:
> The decorator is much better than the variable because it keeps the metadata next to the function.

I don't have a strong opinion about this but I can also see it the other way. Data like which function should be run under the doctest runner can also be better separated from the code itself. Decorators have to actually execute at runtime whereas a static variable like `__doctest_requires__` is just loaded from the .pyc file.

Reply:
> check that `f` is not defined in the doctest unless it is imported

It is still the case that with pytest-doctestplus the function `f` can be used without an explicit import.

That could be changed in pytest-doctestplus but I don't see it as an obstacle to sympy using pytest-doctestplus.

Reply:
Right now this is how all SymPy doctests work. If we change this, then we will be changing the behavior of SymPy doctests. IMO the doctest defaults are bad generally, but they're particularly bad for SymPy where the specific module where a function is defined is usually not something we want users to know or care about. It also doesn't help that we often have a mix of public and non-public functions in the same file. This can lead to all sorts of mistakes because any number of private things could be defined in a file, and the doctest would automatically include them all. And really one of the biggest reasons we do this in SymPy is because almost every doctest uses multiple SymPy functions, not just the one that the doctest is for, because you need different functions to create expressions and so on. So it's much clearer to users where things are coming from, and that different functions are indeed SymPy functions. 

NumPy solves this by adding `np` to the doctest namespace and using `np.` for everything, but we don't really use that sort of convention in the SymPy docs. 

I would say this is a blocker.

Reply:
> I would say this is a blocker.

It is not a blocker for enabling CI tests for pytest-doctestplus so that it is usable for local development. 

We can keep `bin/doctest` and also test it in CI and it will check for this if this particular aspect is important.

I would say though that:

- Other things are significantly more important in terms of docs, doctest, tests, code etc.
- Whatever the problems the proper fix is to fix upstream rather than stay on in-house runners.

Reply:
> > I would say this is a blocker.
> 
> It is not a blocker for enabling CI tests for pytest-doctestplus so that it is usable for local development.

And to be clear this PR does not change anything in CI or affect the current doctest runner. This is not a blocker for this particular PR whose purpose is just to improve running the doctests with pytest-doctestplus for those of us who would prefer to run the tests that way.

Reply:
Sure I agree we can upstream stuff. I almost upstreamed this to pytest once but I never finished it because I couldn't come up with a good name for the option. Personally I think the way our doctest runner works should be the default for tools like `doctest` or `pytest-doctestplus` but I don't have the energy to push for that kind of change.

Reply:
We should discuss this with the `pytest-doctestplus` people. I'm not sure how invasive a change it would be to run the doctests differently like that. Unfortunately `pytest-doctestplus` much like sympy's own doctest runner is still based on monkey-patching the stdlib doctest module which makes things like this potentially more complicated than they would be for a complete standalone runner.

Reply:
>  Unfortunately pytest-doctestplus much like sympy's own doctest runner is still based on monkey-patching the stdlib doctest module which makes things like this potentially more complicated than they would be for a complete standalone runner.

They should just fork it. It's not a bad starting point, but the built-in doctest is kind of a crapshoot whenever you want to do some level of customization. Either the hook you need is there and it's a couple of lines, or it isn't and you have to redefine entire functions just to change one or two things about them. This is why SymPy has such a large amount of "duplicate" code for its doctest runner. There's no special reason why a doctest runner needs to be defined as a subclass of `doctest` classes. 

It has made sense for us because there's still a lot of code in `doctest` that we don't redefine, but for doctestplus I don't see a good reason. Plus, they will invariably end up depending on private methods that will change between Python versions.

Reply:
> They should just fork it.

I agree. My impression is that they have basically adapted sympy's runner and then made it into a pytest plugin as a separate package that can be shared by other projects. That's what we should have done but they already did so we should use it since it is already being maintained and can already be shared.

In making it a separate package though it would be best just to fork doctest.py into that package.

Reply:
And also limit what's exposed in a Python API (if anything is) so that they can change the code architecture if they want to.